File: src/main/java/piles/Pile.java
package piles;

import deck.Card;
import core.InvalidMoveException;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import static core.Constants.WHITE_TEXT;

abstract public class Pile implements Iterable<Card>{
    protected List<Card> cards;
    public Pile(){
        cards = new LinkedList<>();
    }
    public Pile(List<Card> cards){
        this.cards=cards;
    }

    public Card getTopCard(){
        if(!cards.isEmpty()){
            return cards.get(cards.size()-1);
        }
        else{
            return null;
        }
    }
    public void addCard(Card card) throws InvalidMoveException {
        if (this.isCardValid(card)){
            cards.add(card);
        }
        else{
            throw new InvalidMoveException();
        }
    }
    public void removeCard(Card card) throws InvalidMoveException{
        if(!cards.remove(card)) {
            throw new InvalidMoveException();
        }
    }

    public void addCards(List<Card> cards) throws InvalidMoveException {
        for (Card card : cards){
            this.addCard(card);
        }
    }

    public void removeCards(List<Card> cards) throws InvalidMoveException {
        for (Card card : cards){
            this.removeCard(card);
        }
    }

    public void flipTopCard(){
        if(getTopCard()!=null)
            getTopCard().flip();
    }

    public int getSize() {
        return cards.size();
    }

    public String toString() {
        if(getTopCard()==null){
            return WHITE_TEXT +"░░░";
        }
        else{
            return getTopCard().toString();
        }
    }

    public Card getCard(int position) {
        try{
            return cards.get(position);
        }catch (IndexOutOfBoundsException e){
            return null;
        }
    }

    public boolean isEmpty() {
        return cards.isEmpty();
    }

    public abstract boolean isCardValid(Card card);

    @Override
    public Iterator<Card> iterator() {
        return cards.iterator();
    }
}

File: app/src/main/scala/piles/Pile.scala
package piles

import core.Constants.WHITE_TEXT
import core.InvalidMoveException
import deck.Card


abstract class Pile extends Iterable[Card] {
  protected var cards: Seq[Card] = Vector()

  def this(cards: Seq[Card]) = {
    this()
    this.cards = cards
  }

  def getTopCard: Option[Card] = {
    cards.lastOption
  }

  @throws[InvalidMoveException]
  def addCard(card: Card): Unit = {
    if (this.isCardValid(card)) cards = cards :+ card
    else throw new InvalidMoveException
  }

  @throws[InvalidMoveException]
  def removeCard(card: Card): Unit = {
    if (!cards.filter(c => c eq card).contains(card)) throw new InvalidMoveException
    cards = cards.filter(c => !(c eq card))
  }

  @throws[InvalidMoveException]
  def addCards(cards: Seq[Card]): Unit = {
    for (card <- cards) {
      this.addCard(card)
    }
  }

  @throws[InvalidMoveException]
  def removeCards(cards: Seq[Card]): Unit = {
    for (card <- cards) {
      this.removeCard(card)
    }
  }

  def flipTopCard(): Unit = {
    getTopCard match
      case Some(topCard) => topCard.flip()
      case None =>
  }

  def getSize: Int = cards.size

  override def toString: String = {
    getTopCard match
      case Some(topCard) => topCard.toString
      case None => WHITE_TEXT + "░░░"
  }

  def getCard(position: Int): Option[Card] = {
    try Some(cards(position))
    catch {
      case e: IndexOutOfBoundsException =>
        None
    }
  }

  override def isEmpty: Boolean = cards.isEmpty

  def isCardValid(card: Card): Boolean

  override def iterator: Iterator[Card] = cards.iterator

}
